src/app/pages/vm/devices/device-list/device-list/device-list.component.ts

```diff
@@ -0,0 +1,197 @@
+      sortable: true,
+    }),
+    textColumn({
+      title: this.translate.instant('Device'),
+      propertyName: 'dtype',
+      sortable: true,
+      getValue: (device) => {
+        // TODO: getValue is incorrectly typed
+        return this.getDeviceTypeLabel(device) as unknown as VmDevice[keyof VmDevice];
+      },
+    }),
+    textColumn({
+      title: this.translate.instant('Order'),
+      propertyName: 'order',
+      sortable: true,
+    }),
+    actionsColumn({}),
+  ]);
+
+  isLoading$ = new BehaviorSubject<boolean>(true);
+  hasNoData$ = new BehaviorSubject<boolean>(false);
+  hasError$ = new BehaviorSubject<boolean>(false);
+  emptyType$: Observable<EmptyType> = combineLatest([this.isLoading$, this.hasNoData$, this.hasError$]).pipe(
+    switchMap(([isLoading, isNoData, isError]) => {
+      if (isLoading) {
+        return of(EmptyType.Loading);
+      }
+      if (isError) {
+        return of(EmptyType.Errors);
+      }
+      if (isNoData) {
+        return of(EmptyType.NoPageData);
+      }
+      return of(EmptyType.NoSearchResults);
+    }),
+  );
+
+  get emptyConfig(): EmptyService {
+    return this.emptyConfigService;
+  }
+
+  get vmId(): number {
+    return Number(this.route.snapshot.params['pk']);
+  }
+
+  constructor(
+    private ws: WebSocketService,
+    private translate: TranslateService,
+    private slideInService: IxSlideInService,
+    private cdr: ChangeDetectorRef,
+    private emptyConfigService: EmptyService,
+    private matDialog: MatDialog,
+    private route: ActivatedRoute,
+  ) {}
+
+  ngOnInit(): void {
+    this.loadDevices();
+  }
+
+  loadDevices(): void {
+    this.isLoading$.next(true);
+    this.hasError$.next(false);
+
+    this.ws
+      .call('vm.device.query', [[['vm', '=', this.vmId]]])
+      .pipe(untilDestroyed(this))
+      .subscribe({
+        next: (devices) => {
+          this.devices = devices;
+          this.hasNoData$.next(!devices.length);
+          this.dataProvider.setRows(devices);
+          this.isLoading$.next(false);
+          this.setDefaultSort();
+          this.cdr.markForCheck();
+        },
+        error: () => {
+          this.isLoading$.next(false);
+          this.hasError$.next(true);
+          this.cdr.markForCheck();
+        },
+      });
+  }
+
+  onAdd(): void {
+    this.slideInService.open(DeviceFormComponent, {
+      data: {
+        virtualMachineId: this.vmId,
+      },
+    })
+      .slideInClosed$.pipe(untilDestroyed(this))
+      .subscribe(() => {
+        this.loadDevices();
+      });
+  }
+
+  onEdit(device: VmDevice): void {
+    this.slideInService.open(DeviceFormComponent, {
+      data: {
+        device,
+        virtualMachineId: this.vmId,
+      },
+    })
+      .slideInClosed$
+      .pipe(untilDestroyed(this))
+      .subscribe(() => {
+        this.loadDevices();
+      });
+  }
+
+  onDelete(device: VmDevice): void {
+    this.matDialog
+      .open(
+        DeviceDeleteModalComponent,
```

Comment:
I can click `Delete` without Force Delete, maybe it's fine.
<img width="464" alt="Screenshot 2023-10-03 at 17 30 40" src="https://github.com/truenas/webui/assets/22980553/3ed0c60b-3c49-4970-aa07-df5383460f16">